### ISSUE TITLE
Fucntions overwrite each others

### ISSUE THREAD

**Issue by SimonDegraeve** [MatAtBread/nodent#65]
Source
```js
import TaskList from 'listr';

async function getProps(input, options) {
  // Do something async
  return {};
}

export default async function getTaskRunner(input, options = {}, renderer) {
  const props = await getProps(input, options);
  const tasks = [
    {
      title: 'Checking context',
      task: () => new TaskList(getContextTasks(props)),
    },
    {
      title: 'Checking code',
      task: () => new TaskList(getCodeTasks(props)),
    },
    {
      title: 'Releasing',
      task: () => new TaskList(getReleaseTasks(props)),
    },
  ];

  return new TaskList(tasks, { showSubtasks: true, collapse: false, renderer, props });
}
```

Result
```js
'use strict';

exports.__esModule = true;

var _nodentRuntime = require('nodent-runtime');

var _nodentRuntime2 = _interopRequireDefault(_nodentRuntime);

var _listr = require('listr');

var _listr2 = _interopRequireDefault(_listr);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function getProps(input, options) {
  return new Promise(function ($return, $error) {
    // Do something async
    return $return({});
  }.$asyncbind(this));
}

exports['default'] = function getTaskRunner(input) {
  var $args = arguments;
  return new Promise(function ($return, $error) {
    var options, renderer, props, tasks;
    function task() {
      return new _listr2['default'](getContextTasks(props));
    }
    function task() {
      return new _listr2['default'](getCodeTasks(props));
    }
    function task() {
      return new _listr2['default'](getReleaseTasks(props));
    }
    options = $args.length > 1 && $args[1] !== undefined ? $args[1] : {};
    renderer = $args[2];
    return getProps(input, options).then(function ($await_1) {
      props = $await_1;
      tasks = [{
        title: 'Checking context',
        task: task }, {
        title: 'Checking code',
        task: task }, {
        title: 'Releasing',
        task: task }];

      return $return(new _listr2['default'](tasks, { showSubtasks: true, collapse: false, renderer: renderer, props: props }));
    }.$asyncbind(this, $error), $error);
  }.$asyncbind(this));
};

module.exports = exports['default'];
```

I tried with both v2.6.12 and v3.0.1.

You can see that `task` overwrite each others.

**Comment by matAtWork** [comment]
This appears to be (another) nodent/babel issue. Putting the code in the playground (link below) it works. My guess is babel is translating the arrow functions into named functions with the same name that nodent then tries to hoist (which it probably shouldn't do if they have the same name, as it will destory ES6 function expression scope).

Can I ask about the order of plugins? If you have fast-async as the first one, does it work? Meanwhile, I'll look at the hoisting similarly named functions, although it'd be very useful to see what Babel is passing to nodent

http://nodent.mailed.me.uk/#async%20function%20getProps(input%2C%20options)%20%7B%0A%20%20%2F%2F%20Do%20something%20async%0A%20%20return%20%7B%7D%3B%0A%7D%0A%0Aexport%20default%20async%20function%20getTaskRunner(input%2C%20options%20%3D%20%7B%7D%2C%20renderer)%20%7B%0A%20%20const%20props%20%3D%20await%20getProps(input%2C%20options)%3B%0A%20%20const%20tasks%20%3D%20%5B%0A%20%20%20%20%7B%0A%20%20%20%20%20%20title%3A%20'Checking%20context'%2C%0A%20%20%20%20%20%20task%3A%20()%20%3D%3E%20new%20TaskList(getContextTasks(props))%2C%0A%20%20%20%20%7D%2C%0A%20%20%20%20%7B%0A%20%20%20%20%20%20title%3A%20'Checking%20code'%2C%0A%20%20%20%20%20%20task%3A%20()%20%3D%3E%20new%20TaskList(getCodeTasks(props))%2C%0A%20%20%20%20%7D%2C%0A%20%20%20%20%7B%0A%20%20%20%20%20%20title%3A%20'Releasing'%2C%0A%20%20%20%20%20%20task%3A%20()%20%3D%3E%20new%20TaskList(getReleaseTasks(props))%2C%0A%20%20%20%20%7D%2C%0A%20%20%5D%3B%0A%0A%20%20return%20new%20TaskList(tasks%2C%20%7B%20showSubtasks%3A%20true%2C%20collapse%3A%20false%2C%20renderer%2C%20props%20%7D)%3B%0A%7D

**Comment by matAtWork** [comment]
Just checked, and babel-preset-es2015 translates your code to:

```
exports.default = async function getTaskRunner(input) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var renderer = arguments[2];

  var props = await getProps(input, options);
  var tasks = [{
    title: 'Checking context',
    task: function task() {
      return new TaskList(getContextTasks(props));
    }
  }, {
    title: 'Checking code',
    task: function task() {
      return new TaskList(getCodeTasks(props));
    }
  }, {
    title: 'Releasing',
    task: function task() {
      return new TaskList(getReleaseTasks(props));
    }
  }];

  return new TaskList(tasks, { showSubtasks: true, collapse: false, renderer: renderer, props: props });
};
```

There's nothing wrong with the code, but nodent will (incorrectly) hoist the functions. I'll fix that, but if you need a workaround, apply fast-async before es2015.

**Comment by matAtWork** [comment]
Why Babel names the functions, I have no idea - possibly to make debugging a little easier as your functions will have a name. It also assumes your runtime correctly hoists function expressions into the inner function scope, not containing function scope, which is not true of a lot of older browsers.

**Comment by matAtWork** [comment]
Closed by 9f672042cbebfc90fd8d085b7ccec6bc904b3e45 and 64df46ec94f1b1117afb832852c566f8b8563fd4
